trl/trainer/grpo_trainer.py

```diff
@@ -141,13 +168,43 @@ def data_collator(features):  # No data collation is needed in GRPO
         self.max_prompt_length = args.max_prompt_length
         self.max_completion_length = args.max_completion_length  # = |o_i| in the GRPO paper
         self.num_generations = args.num_generations  # = G in the GRPO paper
-        self.generation_config = GenerationConfig(
-            max_new_tokens=self.max_completion_length,
-            do_sample=True,
-            temperature=args.temperature,
-            num_return_sequences=self.num_generations,
-            pad_token_id=processing_class.pad_token_id,
-        )
+        self.use_vllm = args.use_vllm
+
+        if self.use_vllm:
+            if not is_vllm_available():
+                raise ImportError(
+                    "vLLM is not available and `use_vllm` is set to True. Please install vLLM with "
+                    "`pip install vllm` to use it."
+                )
+            self.generation_config = SamplingParams(
+                n=self.num_generations,  # 2 generations per prompt
+                max_tokens=args.max_completion_length,
+                temperature=args.temperature,
+                # top_k=50,
+                # top_p=1.0,
+                detokenize=False,  # to avoid vllm to decode (we don't need it)
+            )
+            # vLLM dynamically adjusts the size of the key-value cache based on available GPU memory at instanciation.
+            # A larger cache size improves speed, so we would expect gpu_memory_utilization=1.
+            # However, at this stage, the optimizer's weights are not yet loaded onto the GPU; they will be loaded
+            # after the first optimizer step and remain in GPU memory throughout training. So we must reserve enough
+            # space for them. Setting gpu_memory_utilization to 0.55 seems to work well in practice.
+            self.llm = LLM(
+                model=model.name_or_path,
```

Comment:
Should we also allow one to pass `revision`?

Reply:
I don't think so, the models weights are loaded from the current model before each generation.

Reply:
Ah perfect, so basically this is handled earlier in the model init?

Reply:
In fact, here you load the model just so that vLLM knows the structure and estimates the amount of memory needed to estimate the amount of memory needed to estimate the amount of memory possible for the KV cache. Before each learning step, the weights of the latest version of the model are loaded:

https://github.com/huggingface/trl/blob/3a280d6749bfbf1fc8c6ccc31d030b1cc97e2d55/trl/trainer/grpo_trainer.py#L267-L269

Revision would only make sense if the structure of the model is not consistent across revisions.

Reply:
Makes sense, thanks!